File: hackerrank/problem solving/algorithms/2 - implementation/37 - jumping on the clouds/oo/jumping_on_the_clouds_oo.cs
// Source: https://www.hackerrank.com/challenges/jumping-on-the-clouds/problem?isFullScreen=true

using System;

public class Solution
{
    public static void Main()
    {
        int _ = _ReadANumber();
        List<int> cloudArray = _ReadAnArray();

        JumpingOnTheClouds obj = new JumpingOnTheClouds(cloudArray);
        int minimumNumberOfJumps = obj.GetMinimumNumberOfJumps();
        Console.WriteLine(minimumNumberOfJumps);
    }

        private static int _ReadANumber()
        {
            int number = int.Parse(Console.ReadLine());
            return number;
        }

        private static List<int> _ReadAnArray()
        {
            List<int> array = Console.ReadLine().Split().Select(int.Parse).ToList();
            return array;
        }
}

    public class JumpingOnTheClouds
    {
        private List<int> _CloudArray;
        private int _MinimumNumberOfJumps;

        public JumpingOnTheClouds(List<int> cloudArray)
        {
            _CloudArray = cloudArray;
            _MinimumNumberOfJumps = 0;

            _CalculateMinimumNumberOfJumps();
        }

            private void _CalculateMinimumNumberOfJumps()
            {
                for (int i = 0, size = _CloudArray.Count - 1; i < size; )
                {
                    i = _GetIndexOfNextJump(i);
                    _MinimumNumberOfJumps++;
                }
            }

                private int _GetIndexOfNextJump(int index)
                {
                    if (_IsNextSecondIndexInsideRange(index + 2))
                        index += _IsNextSecondCloudCumulus(_CloudArray[index + 2]) ? 2 : 1;
                    else
                        index++;
                    return index;
                }

                    private bool _IsNextSecondIndexInsideRange(int index)
                    {
                        return index < _CloudArray.Count;
                    }

                    private bool _IsNextSecondCloudCumulus(int typeOfCloud)
                    {
                        return typeOfCloud == 0;
                    }

        public int GetMinimumNumberOfJumps()
        {
            return _MinimumNumberOfJumps;
        }
    }

File: hackerrank/problem solving/algorithms/2 - implementation/37 - jumping on the clouds/oo/jumping_on_the_clouds_oo.py
# Source: https://www.hackerrank.com/challenges/jumping-on-the-clouds/problem?isFullScreen=true

def main():

    _ = int(input())
    CLOUD_ARRAY = readAnArray()

    OBJ = JumpingOnTheClouds(CLOUD_ARRAY)
    MINIMUM_NUMBER_OF_JUMPS = OBJ.getMinimumNumberOfJumps()
    print(MINIMUM_NUMBER_OF_JUMPS)


def readAnArray():

    ARRAY = list(map(int, input().split()))
    return ARRAY


class JumpingOnTheClouds:

    def __init__(self, cloudArray):

        self.__cloudArray = cloudArray
        self.__minimumNumberOfJumps = 0

        self.__calculateMinimumNumberOfJumps()


    def __calculateMinimumNumberOfJumps(self):

        i = 0
        size = len(self.__cloudArray) - 1

        while i < size:
            i = self.__getIndexOfNextJump(i)
            self.__minimumNumberOfJumps += 1


    def __getIndexOfNextJump(self, index):

        if self.__isNextSecondIndexInsideRange(index + 2):
            index += 2 if self.__isNextSecondCloudCumulus(self.__cloudArray[index + 2]) else 1
        else:
            index += 1
        return index


    def __isNextSecondIndexInsideRange(self, index):

        return index < len(self.__cloudArray)


    def __isNextSecondCloudCumulus(self, typeOfCloud):

        return typeOfCloud == 0


    def getMinimumNumberOfJumps(self):

        return self.__minimumNumberOfJumps


if __name__ == "__main__":
    main()
